core/domain/opportunity_services.py

```diff
@@ -0,0 +1,370 @@
+            exp_ids))
+
+    exploration_opportunity_summary_list = []
+
+    for exp_opportunity_model in exp_opportunity_models:
+        exploration_opportunity_summary = (
+            get_exploration_opportunity_summary_from_model(
+                exp_opportunity_model))
+        exploration_opportunity_summary.story_title = story.title
+        node = story.story_contents.get_node_with_corresponding_exp_id(
+            exploration_opportunity_summary.id)
+        exploration_opportunity_summary.chapter_title = node.title
+        exploration_opportunity_summary.validate()
+
+        exploration_opportunity_summary_list.append(
+            exploration_opportunity_summary)
+
+    _save_multi_exploration_opportunity_summary(
+        exploration_opportunity_summary_list)
+
+
+def delete_exploration_opportunities(exp_ids):
+    """Deletes the exploration_ooportunity_summary models.
+
+    Args:
+        exp_ids: list(str). A list of exploration ids. The opportunity summary
+            model with these ids to be deleted.
+    """
+    exp_opportunity_models = (
+        opportunity_models.ExplorationOpportunitySummaryModel.get_multi(
+            exp_ids))
+    opportunity_models.ExplorationOpportunitySummaryModel.delete_multi(
+        exp_opportunity_models)
+
+
+def update_exploration_opportunities(old_story, new_story):
+    """Updates the opportunities models according to the changes made in the
+    story.
+
+    Args:
+        old_story: Story. The old story object which is now updated.
+        new_story: Story. The new story object.
+    """
+    model_ids_need_update = set([])
+    exp_ids_in_old_story = old_story.story_contents.get_all_linked_exp_ids()
+    exp_ids_in_new_story = new_story.story_contents.get_all_linked_exp_ids()
+    new_added_exp_ids = set(exp_ids_in_new_story) - set(exp_ids_in_old_story)
+    deleted_exp_ids = set(exp_ids_in_old_story) - set(exp_ids_in_new_story)
+    unchanged_exp_ids = set(exp_ids_in_new_story) - new_added_exp_ids
+    if old_story.title != new_story.title:
+        model_ids_need_update |= set(unchanged_exp_ids)
+    else:
+        for exp_id in unchanged_exp_ids:
+            new_node = (
+                new_story.story_contents.get_node_with_corresponding_exp_id(
+                    exp_id))
+            old_node = (
+                old_story.story_contents.get_node_with_corresponding_exp_id(
+                    exp_id))
+            if old_node.title != new_node.title:
+                model_ids_need_update.add(exp_id)
+
+    update_exploration_opportunities_with_new_story(
+        new_story, list(model_ids_need_update))
+    add_new_exploration_opprtunities(new_story, new_added_exp_ids)
+    delete_exploration_opportunities(list(deleted_exp_ids))
+
+
+def get_translation_opportunities(language_code, cursor):
+    """Returns a list of opportunities available for translation in a specific
+    language.
+
+        Args:
```

Comment:
Here and elsewhere (in functions below), deindent this and subsequent lines by 4.